### PR TITLE
Pods 4613

### PR DESCRIPTION
**Pull request by vinnicombe**

### CODE REVIEW

File: app/controllers/financialStatement/ChargeDetailsController.scala [hunk 1]
```diff
@@ -45,42 +48,49 @@ class ChargeDetailsController @Inject()(identify: IdentifierAction,
     with I18nSupport
     with NunjucksSupport {
 
-  def onPageLoad(identifier: String, startDate: LocalDate, chargeReference: String): Action[AnyContent] = identify.async {
+  def onPageLoad(identifier: String, startDate: LocalDate, index: String): Action[AnyContent] = identify.async {
```
Comment (grantrobson) [inline app/controllers/financialStatement/ChargeDetailsController.scala:51]: I'd call this chargeReferenceIndex to make it clear

File: app/controllers/financialStatement/ChargeDetailsController.scala [hunk 2]
```diff
@@ -45,42 +48,49 @@ class ChargeDetailsController @Inject()(identify: IdentifierAction,
...
+          fiCacheConnector.fetch flatMap {
+            case Some(jsValue) =>
+              val chargeRefs: Seq[String] = (jsValue \ "chargeRefs").as[Seq[String]]
+
+              val commonJson = Json.obj(
+                "heading" -> heading(psaFS.filter(_.chargeReference == chargeRefs(index.toInt)).head.chargeType.toString),
+                "isOverdue" -> penaltiesService.isPaymentOverdue(psaFS.filter(_.chargeReference == chargeRefs(index.toInt)).head),
+                "period" -> msg"penalties.period".withArgs(startDate.format(dateFormatterStartDate),
+                  getQuarter(startDate).endDate.format(dateFormatterDMY)),
+                "chargeReference" -> psaFS.filter(_.chargeReference == chargeRefs(index.toInt)).head.chargeReference,
```
Comment (grantrobson) [inline app/controllers/financialStatement/ChargeDetailsController.scala:65]: I'd use headOption and redirect to session expired page if index out of range

File: app/services/paymentsAndCharges/PaymentsAndChargesService.scala [hunk 3]
```diff
@@ -18,89 +18,103 @@ package services.paymentsAndCharges
...
-                messages(s"paymentsAndCharges.visuallyHiddenText", details.chargeReference)
-              ))
+import scala.concurrent.{ExecutionContext, Future}
+
+class PaymentsAndChargesService @Inject()(fiCacheConnector: FinancialInfoCacheConnector) {
+
+  def getPaymentsAndCharges(paymentsAndChargesForAGivenPeriod: Seq[(LocalDate, Seq[SchemeFS])], srn: String)
+                           (implicit messages: Messages, ec: ExecutionContext, hc: HeaderCarrier): Future[Seq[PaymentsAndChargesTable]] = {
+
+    fiCacheConnector.fetch map {
```
Comment (grantrobson) [inline app/services/paymentsAndCharges/PaymentsAndChargesService.scala:46]: This method needs breaking up - it's too long

File: app/controllers/financialStatement/ChargeDetailsController.scala (hunk 1 above)
Comment (vinnicombe) [reply app/controllers/financialStatement/ChargeDetailsController.scala:51]: Done

File: app/controllers/financialStatement/ChargeDetailsController.scala (hunk 2 above)
Comment (vinnicombe) [reply app/controllers/financialStatement/ChargeDetailsController.scala:65]: Done

File: app/services/paymentsAndCharges/PaymentsAndChargesService.scala (hunk 3 above)
Comment (vinnicombe) [reply app/services/paymentsAndCharges/PaymentsAndChargesService.scala:46]: Done

**Pull request merged**